File: apps/node/src/chapter01/kinoko-takenoko.ts
import { readFileSync } from "node:fs";

const BOM = "\uFEFF";

export interface Person {
  height: number;
  weight: number;
  ageGroup: number;
  faction: string;
}

export function loadPeople(csvFile: string): Person[] {
  const text = readFileSync(csvFile, "utf-8");
  const [headerLine = "", ...lines] = (
    text.startsWith(BOM) ? text.slice(BOM.length) : text
  ).split("\n");
  const header = headerLine.split(",");
  const column = (values: string[], name: string): string => {
    const value = values[header.indexOf(name)];
    if (value === undefined) {
      throw new Error(`列 ${name} が見つかりません`);
    }
    return value;
  };
  return lines
    .filter((line) => line.trim() !== "")
    .map((line) => {
      const values = line.split(",");
      return {
        height: Number(column(values, "身長")),
        weight: Number(column(values, "体重")),
        ageGroup: Number(column(values, "年代")),
        faction: column(values, "派閥"),
      };
    });
}

export type Features = Omit<Person, "faction">;

export function splitFeaturesAndLabels(people: Person[]): {
  features: Features[];
  labels: string[];
} {
  return {
    features: people.map(({ height, weight, ageGroup }) => ({
      height,
      weight,
      ageGroup,
    })),
    labels: people.map((person) => person.faction),
  };
}

/** 「20 代ならきのこ派」というルールの年代 */
const KINOKO_AGE_GROUP = 20;

export function predictByRule(features: Features): string {
  return features.ageGroup === KINOKO_AGE_GROUP ? "きのこ" : "たけのこ";
}

export function accuracy(predictions: string[], labels: string[]): number {
  if (predictions.length !== labels.length) {
    throw new Error("予測と正解ラベルの件数が違います");
  }
  const correct = predictions.filter(
    (prediction, i) => prediction === labels[i],
  ).length;
  return correct / labels.length;
}

File: apps/kotlin/src/main/kotlin/chapter01/KinokoTakenoko.kt
package chapter01

import java.io.File

private const val BOM = '\uFEFF'

data class Person(
    val height: Int,
    val weight: Int,
    val ageGroup: Int,
    val faction: String,
)

fun loadPeople(csvFile: File): List<Person> {
    val lines = csvFile.readLines(Charsets.UTF_8)
    val header = lines.first().removePrefix(BOM.toString()).split(",")
    val index = header.withIndex().associate { (i, name) -> name to i }
    return lines.drop(1).filter { it.isNotBlank() }.map { line ->
        val values = line.split(",")
        Person(
            height = values[index.getValue("身長")].toInt(),
            weight = values[index.getValue("体重")].toInt(),
            ageGroup = values[index.getValue("年代")].toInt(),
            faction = values[index.getValue("派閥")],
        )
    }
}

data class Features(
    val height: Int,
    val weight: Int,
    val ageGroup: Int,
)

fun splitFeaturesAndLabels(people: List<Person>): Pair<List<Features>, List<String>> {
    val features = people.map { Features(it.height, it.weight, it.ageGroup) }
    val labels = people.map { it.faction }
    return features to labels
}

/** 「20 代ならきのこ派」というルールの年代 */
private const val KINOKO_AGE_GROUP = 20

fun predictByRule(features: Features): String = if (features.ageGroup == KINOKO_AGE_GROUP) "きのこ" else "たけのこ"

fun accuracy(
    predictions: List<String>,
    labels: List<String>,
): Double {
    require(predictions.size == labels.size) { "予測と正解ラベルの件数が違います" }
    val correct = predictions.zip(labels).count { (p, t) -> p == t }
    return correct.toDouble() / labels.size
}
